scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/vfl/algorithm/PSI.scala

```diff
@@ -18,38 +18,61 @@ package com.intel.analytics.bigdl.ppml.vfl
 
 import java.util
 
-import com.intel.analytics.bigdl.ppml.psi.test.TestUtils
+import com.intel.analytics.bigdl.ppml.psi.HashingUtils
 import com.intel.analytics.bigdl.ppml.vfl.utils.FLClientClosable
 import org.apache.log4j.Logger
+import scala.util.control.Breaks._
 
 import scala.collection.JavaConverters._
 
 class PSI() extends FLClientClosable {
   val logger = Logger.getLogger(getClass)
+  var salt: String = null
   private var hashedKeyPairs: Map[String, String] = null
   def getHashedKeyPairs() = {
     hashedKeyPairs
   }
   private def uploadKeys(keys: Array[String]) = {
     val salt = getSalt
     logger.debug("Client get Salt=" + salt)
-    val hashedKeys = TestUtils.parallelToSHAHexString(keys, salt)
+    val hashedKeys = HashingUtils.parallelToSHAHexString(keys, salt)
     hashedKeyPairs = hashedKeys.zip(keys).toMap
     // Hash(IDs, salt) into hashed IDs
     logger.debug("HashedIDs Size = " + hashedKeys.size)
     uploadSet(hashedKeys.toList.asJava)
 
   }
 
-  def getSalt(): String = flClient.psiStub.getSalt()
-  def getSalt(name: String, clientNum: Int, secureCode: String): String =
-    flClient.psiStub.getSalt(name, clientNum, secureCode)
+  def getSalt(): String = {
+    salt = flClient.psiStub.getSalt()
+    salt
+  }
+  def getSalt(name: String, clientNum: Int, secureCode: String): String = {
+    salt = flClient.psiStub.getSalt(name, clientNum, secureCode)
+    salt
+  }
 
-  def uploadSet(hashedIdArray: util.List[String]): Unit = {
+  def uploadSet(ids: util.List[String]): Unit = {
+    val hashedIdArray = HashingUtils.parallelToSHAHexString(ids, salt)
     flClient.psiStub.uploadSet(hashedIdArray)
   }
 
-  def downloadIntersection(): util.List[String] = flClient.psiStub.downloadIntersection
-
+  def downloadIntersection(max_try: Int = 20, retry: Long = 1000): util.List[String] = {
+    var intersection: util.List[String] = null
+    breakable {
+      for (i <- 0 until max_try) {
+        intersection = flClient.psiStub.downloadIntersection
+        if (intersection == null) {
+          logger.info(s"Got empty intersection, retry in $retry ms")
+          Thread.sleep(retry)
+        }
+        else {
+          logger.info("Intersection successful. Intersection's size is " + intersection.size + ".")
+          break
+        }
+      }
+    }
+    intersection
```

Comment:
report error if `intersection` is null? shall we make `downloadIntersection` blocking at the server side?

Reply:
Current `downloadIntersection` is blocked at server side using 
```
if(null == intersection) {
   this.wait();
}
```
Maybe it could just be replaced `if` by `while` then multiple-try is not necessary?